modules/navier_stokes/doc/content/source/kernels/INSADMomentumCoupledForce.md

```diff
@@ -0,0 +1,13 @@
+# INSADMomentumCoupledForce
```

Comment:
It seems there is little overlap of this one and body force term? I wonder if there is any particular case to use this term.

Reply:
What body force term?

Reply:
For example, gravity? if the function is 0 0 -0.98 then it is the same as gravity term, correct?

Reply:
Hmm you are indeed right that this object has a superset of the capability provided by `INSADGravityForce`. I'm trying to think how I would generalize the design. A user could add multiple `INSADMomentumCoupledForce` objects in their input file. Each would have to use a unique coupled variable or unique function, and then I'd have to create an arbitrary number of coupled force properties in the `INSADMaterial`. Conceptually that seems straightforward enough. More difficult and painstaking would be adding this capability to the new `INSAction`. It would have a parameter like `params.addParam<std::vector<VariableName>>("additional_coupled_variable_forces")` and `params.addParam<std::vector<FunctionName>>("additional_coupled_function_forces")` and then parse those parameters and add the objects.

This would probably be 1-2 days work. Not sure if I have the time for that right now unless Griffin cares about it.

Reply:
I see it becomes a little messy for Action. I would not worry about this for now and we can improve it in the future.

Reply:
Actually I could still do this with a single object. I would just change the coupled variables/functions parameters to take multiple arguments (the coupled variables parameter already does this anyways)

Reply:
Hmm @jiangwen84 I think there's probably still value in having the `INSADGravityForce` object. For example in my first naive attempt to simply sub the `INSADMomentumCoupledForce` in the boussinesq test, I applied this patch:
```
diff --git a/modules/navier_stokes/test/tests/ins/boussinesq/boussinesq_square.i b/modules/navier_stokes/test/tests/ins/boussinesq/boussinesq_square.i
index 24c90922f8..8554a00be1 100644
--- a/modules/navier_stokes/test/tests/ins/boussinesq/boussinesq_square.i
+++ b/modules/navier_stokes/test/tests/ins/boussinesq/boussinesq_square.i
@@ -126,9 +126,9 @@
     gravity = '0 -9.81 0'
   [../]
   [./gravity]
-    type = INSADGravityForce
+    type = INSADMomentumCoupledForce
     variable = velocity
-    gravity = '0 -9.81 0'
+    vector_function = 'vector_func'
   [../]
 []
 
@@ -151,3 +151,11 @@
     temperature = temp
   []
 []
+
+[Functions]
+  [vector_func]
+    type = ParsedVectorFunction
+    value_x = '0'
+    value_y = '-9.81'
+  []
+[]
```
Result: exodiff. The reason being that the gravity vector is multiplied by rho.

Reply:
Sounds good to me. Thanks!